### DISCUSSION TITLE
Beacon Chain alternative: PoS over PoA (Copied from Team discussion)

### DISCUSSION THREAD

**Discussion by hewison-chris** [bosagora/AgoraMainNet#3] [Ideas]
@[omerfirmak](https://github.com/omerfirmak)
omerfirmak
[22 hours ago ](https://github.com/orgs/zeroone-boa/teams/zeroone/discussions/2)• 
Authority-Round (AuRa) is a PoA variant consensus algorithm, which I believe could be a plausible alternative, for our use case, to Beacon chain.

[AuRa](https://openethereum.github.io/Aura) enables a set of trusted nodes to take turns proposing blocks and requires majority of authorities to sign-off on the proposed block for finalization. In that sense, it is really similar to SCP that we are really familiar with.

But it is much more powerful in terms of being configurable. It uses smart contracts to let developers modify the network's behavior.

Most importantly, it lets the validator set to be controlled by a [smart contract](https://openethereum.github.io/Validator-Set), so we can dynamically change the validator set. We could easily deploy a contract that would accept 40K BOAs to include a node to the validator set.

contract ReportingValidatorSet {
    // all same as ValidatorSet
    event InitiateChange(bytes32 indexed _parent_hash, address[] _new_set);

    function getValidators() constant returns (address[] _validators);
    function finalizeChange();

    // Reporting functions: operate on current validator set.
    // malicious behavior requires proof, which will vary by engine.

    function reportBenign(address validator, uint256 blockNumber);
    function reportMalicious(address validator, uint256 blockNumber, bytes proof);
}
AuRa engine also supports automatically reporting misbehaving nodes, according to the engine rules, This would allow us to slash a validator if needed. So, in practice we would be running a set of validators controlled by a deposit contract running PoA for reaching consensus.

Secondly, block rewards could be set by a [smart contract](https://openethereum.github.io/Block-Reward-Contract) as well.

interface BlockReward {
    // produce rewards for the given benefactors,
    // with corresponding reward codes.
    // only callable by `SYSTEM_ADDRESS`
    function reward(address[] benefactors, uint16[] kind)
        external
        returns (address[], uint256[]);
}
Commons budget and any other complex rewarding scheme could be implemented in this contract with this really simple ABI.

There are multiple advantages of going with AuRa instead of Beacon chain. To list a few;

We would only be writing our own smart contracts and the Ethereum client code would be abstracted away from us, greatly reducing the development time.
Decouples us from the Eth 2.0 timeline. ETH2.0 is still not in production and yet to prove itself.
AuRa has multiple implementations in the market, we could pick one or use bunch of them together. So we wouldnt be tied to a single implementation.
Beacon protocol requires thousands of nodes to be available. This would put immense load on operation budget of foundation for the mainnet. With AuRa we could run very few foundation validators if we want. With fewer nodes, maintaining the mainnet would be much easier as well.
Beacon protocol is too complex for our needs at this time, this will slow down the development. Sharding, crosslinks, layered execution and consensus etc.
Long story short, with AuRa we will get a network really similar to what we aimed with original Agora and SCP, in terms of how consensus behaves. but have smart contract abilities on top of that.

Mature AuRa implementations;

https://github.com/NethermindEth/nethermind
https://github.com/openethereum/openethereum
 @[MichaelKim20](https://github.com/MichaelKim20)
MichaelKim20 [6 hours ago](https://github.com/orgs/zeroone-boa/teams/zeroone/discussions/2/comments/1) • 
I made the issue of github. I want the team members to divide these issues and proceed.
https://github.com/zeroone-boa/agoranet/issues/1

 @[MichaelKim20](https://github.com/MichaelKim20)
MichaelKim20 [4 hours ago](https://github.com/orgs/zeroone-boa/teams/zeroone/discussions/2/comments/2) • 
about Authority-Round (AuRa)
The main issues to be looked at in this are as follows.
Is there a protocol that allows the validator to register or unregister its role as a validator?
Is the amount of the steak fixed? If fixed, should the POA node run each time more validators are performed??
If the staking amount is not fixed, is there a priority for voting rights during block verification?
How many validator nodes can I run?
If the staking amount is fixed, more than 1000 pieces will have to proceed with POA.
Is this possible with POA?
I insist on using Ethereum's beacon if it is implemented and cannot be told immediately.
This is because major issues of POS have already been resolved in Ethereum's beacon chain.

 @[hewison-chris](https://github.com/hewison-chris)
hewison-chris [3 hours ago](https://github.com/orgs/zeroone-boa/teams/zeroone/discussions/2/comments/3)
Let's keep the tasks as they are but look at what changes would be required for Beacon vs AuRa.
I think all of the concerns you mention above can be achieved using the ValidatorSet and Block Rewards Smart Contracts plus a few new contracts.
Staking (Freezing funds of 40K Boa) / register to be a validator and de-register (un-freezing) would just be done in Smart Contracts.

 @[omerfirmak](https://github.com/omerfirmak)
omerfirmak [2 hours ago](https://github.com/orgs/zeroone-boa/teams/zeroone/discussions/2/comments/4) Author
@MichaelKim20, please read thru the docs I linked in the first post. To help everyone get a better understanding let me quickly outline some of the contracts we might have in the system;

Deposit Contract

It would let users deposit and withdraw their BOAs to adjust their validator status. How much they should deposit would only be limited by a minimum amount, they could decide to deposit more than that. Once a pubkey's validator status changes (either becomes or stops being eligible), this contract would invoke ValidatorSet contract to change the validators.

ValidatorSet

This would probably just be a simple set data structure implementation and will provide add/remove interfaces for only deposit contract to call and an interface for system to get all currently active validators.

It might also implement reportBenign and reportMaliciousfor slashing.

Block Rewards

This contract will implement rewards that go to the validators and commons budget. It could query ValidatorSet to get the active validator set and Deposit Contract to get amount of their staked BOAs, so that the rewards could be distributed accordingly.

Commons Budget

Commons budget would have to be controlled by a smart contract too, probably only be allowed to be called by the Votera smart contract.

Votera

Will handle congress voting, it can query ValidatorSet to get which pubkeys are allowed to vote and once a proposal passes, it would call Commons Budget contract if needed to release some funds from it.

**Comment by hewison-chris** [comment]
Next step:
1) As the Validator set is controlled by the smart contract there should be no limit to the number of validators. However we should check that if it is possible to run any number of validators with AuRa consensus.
2) Which tools (equivalent to https://github.com/ethereum/staking-deposit-cli and https://launchpad.ethereum.org) are available or how easily can we  create tools to manage the staking and enrolling.
3) Run a POC with some validators and if possible a client capable of enrolling showing the state of the validators.

**Reply by MichaelKim20** [reply]
Ethereum Beacon randomly elects 128 committees from many validators to create a single block. Therefore, even if the number of validators increases, the total number of validators does not affect the agreement because the number of validators participating in the agreement is always the same.
What about AuRa? If all nodes have to participate in the creation of each block, the agreement may not go smoothly as the number of nodes increases.
Therefore, I would like to receive the result of executing more than 500 validators. I can trust this after I see it.

**Reply by MichaelKim20** [reply]
This will work fine in small dense networks, but I think many issues will arise in systems where multiple nodes are widely distributed.
First, can someone tell me the Aura consensus algorithm in detail?
I couldn't find out what it was different from POA.

**Comment by MichaelKim20** [comment]
I think Aura can register and unregister the validator, and reward and slash.
However, it is doubtful whether this can be used by multiple nodes and validators.

https://www.xdaichain.com/ only operates 19 validators.
Please prove that it is a blockchain that can be operated by many validator

I should focus on Ethereum beacon chain before then.

**Comment by MichaelKim20** [comment]
@hewison-chris  
Here's my opinion on this part you described

Beacon protocol requires thousands of nodes to be available. This would put immense load on operation budget of foundation for the mainnet. With AuRa we could run very few foundation validators if we want. With fewer nodes, maintaining the mainnet would be much easier as well.

We should consider open membership.
We are not offering a small number of private blockchains because of the foundation's cost.
We need to create a public blockchain.
And I know that beacons can run multiple validators on one node.

**Comment by hewison-chris** [comment]
[POSDAO - Proof of Stake Decentralized Autonomous Organization (15).pdf](https://github.com/zeroone-boa/AgoraMainNet/files/8649712/POSDAO.-.Proof.of.Stake.Decentralized.Autonomous.Organization.15.pdf)
After reading this I found out that the maximum validators set to 19 is the limit to the set that are selected from the `validator set` 
 (`Validator candidates`) to validate the current Epoch. There is no limit on the `Enrolled validators` just to how many are used to generate the next epoch of blocks. This is similar to Beacon where 256 are selected from the total number.

**Comment by omerfirmak** [comment]
Please keep in mind that using AuRa at this point to get CoinNet released faster does not keep us from switching to Beacon chain once we have enough time and resources. 

The beacon chain being more scalable does not really serve us much of a purpose when you consider that we only had a **single** person running a validator in Agora testnet outside of our development team.

**Reply by MichaelKim20** [reply]
The network currently in operation has only 19 validators. I believe that we should run at least 1000 nodes and decide

**Reply by omerfirmak** [reply]
see https://github.com/orgs/zeroone-boa/discussions/3#discussioncomment-2712499

**Reply by MichaelKim20** [reply]
I've read it. Why are there only 19 validators in their network?

**Reply by hewison-chris** [reply]
They have up to 3000 validators which they call `candidate validators` and 19 which are slected to be active for the next Epoch.

**Reply by MichaelKim20** [reply]
https://blockscout.com/xdai/mainnet/validators

**Reply by omerfirmak** [reply]
it is actually exactly the same to what beacon does. They have a large pool of candidate validators and they pick a subset of them to become active on rotation.

**Comment by MichaelKim20** [comment]
Is the amount of steak fixed?

**Reply by MichaelKim20** [reply]
Or does the validator participation rate adjust according to the amount of steak?

**Reply by omerfirmak** [reply]
that is completely up to us, we will implement it in the smart contract

**Reply by hewison-chris** [reply]
The steak is not fixed and the more staked on a `candidate validator` the more likely it is chosen to be a Validator for the next `Epoch`. See sections 1.3 for overview and 9.1 for details of Smart Contracts that define the rules in [POSDAO - Proof of Stake Decentralized Autonomous Organization (15).pdf](https://github.com/zeroone-boa/AgoraMainNet/files/8649712/POSDAO.-.Proof.of.Stake.Decentralized.Autonomous.Organization.15.pdf)

**Comment by MichaelKim20** [comment]
I wanted to confirm this.
1. When the staking amount is fixed, the number of nodes increases and the network load increases.
2. If the staking amount is not fixed, when the validators are randomly selected, the staking amount should be selected more.

I found out from the white paper that POSDAO is more selected as a validator if the staking amount is high. I like this one.

**Comment by MichaelKim20** [comment]
Are you sure the implementation they use is `openethereum`?

**Reply by omerfirmak** [reply]
they use openethereum and nethermind, we should do the same and use a mix of those on foundation nodes.

**Comment by MichaelKim20** [comment]
I think they are also changing to Ethereum's beacon chain. What should we think of this?
https://beacon.gnosischain.com/
https://gnosischain.gitbook.io/gnosis-beacon-chain/

**Reply by omerfirmak** [reply]
AuRa is, more or less, simplified beacon chain. It will allow us to release faster, then we can build beacon chain on top of it.

**Reply by MichaelKim20** [reply]
Should we now copy what they gave up? I'm worried about what other people will think.

**Reply by omerfirmak** [reply]
They ran this setup for ages, they didnt give up. they are just moving forward.
Other people will think worse of us the more we delay the coinnet release.

**Reply by MichaelKim20** [reply]
https://gnosischain.gitbook.io/gnosis-beacon-chain/#gnosis-chain
Gnosis Chain will continue to run in its current form  until the merge with the Gnosis Beacon Chain.
This merge will follow a similar trajectory to the Ethereum merge.

```
Gnosis Chain (formerly the xDai Chain) is the associated execution-layer 
EVM chain for stable transactions. Gnosis Chain uses the xDai token for 
transactions and includes a wide-ranging group of projects and users.
Much like Ethereum, Gnosis Chain will continue to run in its current form 
until the merge with the Gnosis Beacon Chain. 
This merge will follow a similar trajectory to the Ethereum merge.
After the merge, validators (stakers) on the GBC will begin validating transactions 
on the Gnosis Chain; the Gnosis Chain will become a shard within the larger GBC infrastructure.
Much more information on Gnosis Chain, usage, and applications is available on the 
[GC site](https://www.xdaichain.com/). In addition, 
here are a few quick links to GC-relevant resources.
```

**Reply by MichaelKim20** [reply]
I never asked to delay the launch of CoinNet. As planned, I asked to do it in October. And it's important to choose a definite solution and it's not you who want to change what's already been decided?
